### PR TITLE
[MAINT] move surface IO module to stable surface API

### LINKED ISSUE #4568: Move experimental Surface related API into stable codebase
- [x] move and rename pytest fixture from experimental/conftest.py - see #4638 
- [x] Update experimental surface example
  - [x] plot_surf_atlas - #4678
  - [x] plot_localizer_surface_analysis - #4681
  - [x] plot_3d_map_to_surface_projection - #4674
  - [x] plot_surface_bids_analysis - #4682
  - [x] plot_surface_image_and_maskers - #4684
  - [x] plot_surf_stat_map - #4683
- [ ] Move modules:
  - [x] IO functions - #4685
  - [x] plotting functions - #4688  
  - [x] maskers - #4692
  - [x] label masker #4714 
  - [ ] datasets - #4693 
  - [x] surface - #4723
- [ ] Check public class/method and function names
- [ ] Update doc/modules/surface.rst and remove experimental.rst
- [ ] Increase testing
- [x] What to deprecate from old surface.py API?
- [ ] Check/improve docstrings (still need review of course)

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #4568 

Changes proposed in this pull request:

- move mesh_to_gifti and data_to_gifti to stable API
- use stable code to load data and mesh
- remove experimental io module

### CODE REVIEW

File: nilearn/surface/tests/test_surface.py [hunk 1]
```diff
@@ -71,6 +71,18 @@ def data(self):
         return self._data
 
 
+def test_load_surf_data(tmp_path):
```
Comment (Remi-Gau) [inline nilearn/surface/tests/test_surface.py:74]: not sure we should keep this given that we already have tests for this function

Comment (bthirion) [reply nilearn/surface/tests/test_surface.py:74]: Agreed.

File: nilearn/surface/surface.py [hunk 2]
```diff
@@ -1237,3 +1237,50 @@ def check_surface(surface):
...
     mesh, data = check_mesh_and_data(surface.mesh, surface.data)
     return Surface(mesh, data)
+
+
+def mesh_to_gifti(
+    coordinates,
+    faces,
+    gifti_file,
+):
+    """Write surface mesh to gifti file on disk."""
```
Comment (bthirion) [inline nilearn/surface/surface.py:1247]: If this is meant to be part of the public API, the docstring should be more complete

File: nilearn/surface/surface.py [hunk 3]
```diff
@@ -1237,3 +1237,50 @@ def check_surface(surface):
...
+    faces_array = gifti.GiftiDataArray(
+        faces, intent="NIFTI_INTENT_TRIANGLE", datatype="int32"
+    )
+    gifti_img.add_gifti_data_array(coords_array)
+    gifti_img.add_gifti_data_array(faces_array)
+    gifti_img.to_filename(gifti_file)
+
+
+def data_to_gifti(data, gifti_file):
+    """Save data from Polydata to a gifti file."""
```
Comment (bthirion) [inline nilearn/surface/surface.py:1262]: Same thing here

**Review by bthirion — COMMENTED** [review]
Just a point on docstrings.

File: nilearn/surface/surface.py (hunk 2 above)
Comment (Remi-Gau) [reply nilearn/surface/surface.py:1247]: yeah
I was thinking of doing a global pass to update them all at the end but probably better to do it as we integrate the new code.

Comment (Remi-Gau) [reply nilearn/surface/surface.py]: should be better now

File: nilearn/surface/surface.py (hunk 3 above)
Comment (Remi-Gau) [reply nilearn/surface/surface.py]: should be better now

File: nilearn/surface/tests/test_surface.py [hunk 4]
```diff
@@ -437,20 +436,9 @@ def test_load_surf_data_file_glob(tmp_path):
         load_surf_data(tmp_path / "*.gii")
 
 
-def _flat_mesh(x_s, y_s, z=0):
```
Comment (Remi-Gau) [inline nilearn/surface/tests/test_surface.py:440]: duplicate from nilearn.surface._testing

**Review by bthirion — COMMENTED** [review]
There are several codecov complaints on that one.

File: nilearn/conftest.py [hunk 5]
```diff
@@ -380,6 +380,15 @@ def img_atlas(shape_3d_default, affine_mni):
 
 
 # ------------------------ SURFACE ------------------------#
+@pytest.fixture
+def single_mesh(rng):
```
Comment (Remi-Gau) [inline nilearn/conftest.py:384]: this will replace the `generate_surf` from nilearn.surface.tests._testing

File: nilearn/conftest.py [hunk 6]
```diff
@@ -380,6 +380,15 @@ def img_atlas(shape_3d_default, affine_mni):
...
+    """Create random coordinates and faces for a single mesh.
+
+    This does not generate meaningful surfaces.
```
Comment (bthirion) [inline nilearn/conftest.py:387]: Wouldn't it be better to generate a meanginful surface ?

**Review by bthirion — COMMENTED** [review]
Looks almost good to me.

File: nilearn/conftest.py (hunk 6 above)
Comment (Remi-Gau) [reply nilearn/conftest.py:387]: A the moment this does exactly the same as what this function does got other tests.

https://github.com/nilearn/nilearn/blob/c984ebe0b52c6cfc9cf7d5f33979250f05866767/nilearn/surface/tests/_testing.py#L9-L18

more refactoring of the test fixtures is required for the surface part:
see https://github.com/nilearn/nilearn/issues/4686

but i'd prefer to keep that for another PR

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Pull request merged**